README.md
# ack-host
Common node hosting functionality to help reduce deployment of repetitious server functionality

### Table of Contents
- [Examples](#examples)
  - [Simple Example](#simple-example)
  - [Robust Example](#robust-example)
- [Documentation](#documentation)
  - [Host](#host)
  - [API](#api)
  - [Website](#website)
  - [Express](#express)
  - [Built-in Middleware](#built-in-middleware)
  - [Route Metadata](#route-metadata)
  - [Testing](#testing)

## Examples

### Simple Example
```javascript
var path = require('path')
var ackHost = require('ack-host')
var options = require('./ssl-options-file.json')//adds ssl keys to options

// create http site that relocates to https
ackHost.host(80, 'webapp.com').secure()

// create https server that will serve static files on www path or relocates to www path
ackHost.host(443, 'webapp.com', options)
.static('www', path.join(__dirname,'www'))
.relocate('index.html','www/')// relocate all other requests to an index

//start hosting two ports 80 and 443
ackHost.start()
.then(config=>console.log('started on ports', Object.keys(config)))
.catch(e=>console.error('Coult Not Start Server',e))
```

### Robust Example
```javascript
var path = require('path')
var ackHost = require('ack-host')
var options = require('./ssl-options-file.json')//adds ssl keys to options

// create http server site relocates to https
ackHost.host(80, 'webapp.com').secure()

ackHost.host(443, 'webapp.com', options)
.logging()
.noRobots()
.timeout(2000)
.cors()
.static('www', path.join(__dirname,'www'))
.ignoreFavors()
.compress()
.logging('/admin', {stream:process.stdout})
.localNetworkOnly('/admin/index.html')
.respond('/admin/index.html', 'You have come to the admin place')
.closeProductionErrors('/admin')
.consoleNonProductionErrors('/dev')
.throw('/test-error-throw', 'Not a place you should be!')
.relocate('/toGoogle', 'http://google.com')
.respond('/index.html','you have come to the right place')
.relocate('/index.html')

//start hosting two ports 80 and 443
ackHost.start()
.then(config=>console.log('started on ports', Object.keys(config)))
.catch(e=>console.error('Coult Not Start Server',e))
```



## Documentation

### Host
Most barebones provided method of http hosting. Barely any server assumptions. Registers and returns [connect app](https://www.npmjs.com/package/connect)
```javascript
const app = ackHost.host(port, host, options)
```

- Arguments
  - **port** Number|Array-of-Numbers - port or ports to listen on
  - **host** String|Arrray-of-String - host names to listen for requests
  - **options** 
    - **cert** String - for ssl
    - **key** String - for ssl
    - **SNICallback** Function - for ssl


### API
Assumes timeout of 10 seconds. A 404 will be thrown when no route is matched. Registers and returns [express app](https://www.npmjs.com/package/express)
```javascript
const app = ackHost.api(port, host, options)
```

- Arguments
  - **port** Number|Array-of-Numbers - port or ports to listen on
  - **host** String|Arrray-of-String - host names to listen for requests
  - **options** 
    - **timeout** Number = 10000
    - **cert** String - for ssl
    - **key** String - for ssl
    - **SNICallback** Function - for ssl


### Website
Assumes timeout of 30 seconds. A 404 will be thrown when no route is matched. Client input such as form data, is always pre-parsed. Registers and returns [express app](https://www.npmjs.com/package/express)
```javascript
const app = ackHost.website(port, host, options)
```

- Arguments
  - **port** Number|Array-of-Numbers - port or ports to listen on
  - **host** String|Arrray-of-String - host names to listen for requests
  - **options** 
    - **timeout** Number = 10000
    - **cert** String - for ssl
    - **key** String - for ssl
    - **SNICallback** Function - for ssl

### Express
Registers and returns [express app](https://www.npmjs.com/package/express)
```javascript
const app = ackHost.express(port, host, options)
```

- Arguments
  - **port** Number|Array-of-Numbers - port or ports to listen on
  - **host** String|Arrray-of-String - host names to listen for requests
  - **options** 
    - **cert** String - for ssl
    - **key** String - for ssl
    - **SNICallback** Function - for ssl

### start
All sites registered with ack-host, will now be started
```javascript
require('ack-host')
.start(port=>console.log('port started:',port))
.then(config=>console.log('all ports started:',Object.keys(config)))
```

### startOnePort
Great for forcing server to start on first available port
```javascript
require('ack-host')
.host([8080,8081,8082])
.startOnePort(port=>console.log('Found open port:',port))
.then(config=>console.log('only one port started'))
```

### Built-in Middleware
Using on of the hosting methods above (host/api/website/express), you can invoke any of the following functionality.

#### .secure
Relocate http to https
```javascript
app.secure()
```

#### .noRobots
WHEN route matches /\/robots\.txt$/ THEN responds with text/plain message of "User-agent: *\rDisallow: /"
```javascript
app.noRobots(2000)
```

#### .timeout
request timeout
```javascript
app.timeout(2000)
```

#### .static
serve static client files in directory named assets
```javascript
app.static('www', path.join(__dirname,'www'))
```

#### .cors
enable cross origin requests from all domains
```javascript
app.cors()
```

#### .ignoreFavors
ignore favicon like requests
```javascript
app.ignoreFavors()
```

#### .compress
gzip responses where possible
```javascript
app.compress()
```

#### .localNetworkOnly
Only allow local area network requests to /admin path
```javascript
app.localNetworkOnly('/admin')
```

#### .logging
Log all requests to /admin path. Stream option not required and auto defaults to console.log aka process.stdout
```javascript
app.logging('/admin', {stream:process.stdout})
```

#### .closeProductionErrors
if code breaks in admin path respond 500 with little info
```javascript
app.closeProductionErrors('/admin')
```

### .consoleNonProductionErrors
if code breaks in dev path respond 500 with a lot of info
```javascript
app.consoleNonProductionErrors('/dev')
```

### .throw
throw 500 with custom message to /xxx path requests
```javascript
app.throw('/xxx', 'Not a place you should be!')
```

### .relocate
relocate one path
```javascript
app.relocate('/toGoogle', 'http://google.com')
```

relocate all other requests to an index
```javascript
app.relocate('index.html')
```

## Route Metadata
Create route explanation and defintions using GET, POST, PUT, DELETE or USE metadata.

### .meta()
- **details** String - route descriptive details
- **sample**
  - **params** Object - Key value pair to fill in path :params
  - **request** String|Object|Array|Function - request body to be used with sample test. When Function, the function is called for its request body data
  - **response** String|Function-Returns-String - response body to be used to fake a server response during sample testing
  - **test** Boolean|Object
    - **cases** Array-of-Functions|Function(response,assert)
    - **only** Boolean
    - **skip** Boolean
    - **timeout** Number = 2000 - how long to wait before test is aborted

Route Metadata Example
```javascript
const ackHost = require('ack-host')
const port = 3000

//create arbitrary server
const app = ackHost.api(port).logging().timeout(2000).cors()

app.get('/index.html', ackHost.router.respond('you have come to the right place'))
.meta({
  details:"main page of our website",
  sample:()=>({
    test:{
      only:true,
      cases:[(response,assert)=>{
        assert.equal(response.body, 'you have come to the right place')
      }]
    }
  })
})

app.get('/admin/index.html', ackHost.router.respond('you have come to the admin'))
.meta({
  details:"admin page of our website",
  sample:()=>({
    test:{skip:true},
    cases:(response,assert)=>{
      assert.equal(response.body, 'you have come to the right place')
    }
  })
})

app.get('/:id/data.json', (req,res)=>res.json({hello:'world '+req.params.id}))
.meta({
  details:"main page of our website",
  sample:()=>({
    test:{
      only:false, skip:false,
      params:{id:234},
      cases:[(response,assert)=>{
        assert.equal(response.body, '{"hello":"world 234"}')
      }]
    }
  })
})
```


## Testing
Test server apps registered through ack-host

### Testing Options
- port   : Number = all-ports
- host   : String = localhost
- method : String (GET|POST|PUT|DELETE)
- logTo  : Object = console.log - not yet implemented

### Large Test Example
```javascript
const ackHost = require('ack-host')
const port = 3000

//create arbitrary server
const app = ackHost.api(port).logging().timeout(2000).cors()

app.use(/\/robots\.txt/, ackHost.router.noRobots())
.meta({
  details:"prevents robots from crawling our site",
  sample:()=>({
    path:"/robots.txt",
    test:true
  })
})

app.get('/index.html', ackHost.router.respond('you have come to the right place'))
.meta({
  details:"main page of our website",
  sample:()=>({
    test:true
  })
})

app.use( ackHost.router.relocate('/index.html') )
.meta({
  details:"redirects users to our index page",
  sample:()=>({
    test:true
  })
})

ackHost.startAndTest()
.then(results=>{
  console.log('Finished Testing Ports:', Object.keys(results))
  
  for(var port in results){
    console.log('Port Passing', port, results[port].passing.length)
    console.log('Port Failing', port, results[port].failing.length)
  }
})
.catch(e=>console.error('Coult Not Start and Test Server',e))
```

### Start Server and Test
Will start ack-host server and then test all defined server apps
```javascript
ackHost.startAndTest(options)
```

### Test Server Already Running
Will test all defined apps against already running server
```javascript
ackHost.test(options)
```

### Test Server Already Running With Limited Routes by Apps
Will test provided apps against already running server
```javascript
ackHost.testApps(apps,options)
```

CHANGELOG.md
# ack-host - Change Log
All notable changes to this project will be documented here.

## [0.8.0] - 2019-05-26
- TypeScripted

## [0.7.0] - 2019-02-05
- updated packages and security fixes

## [0.6.2] - 2018-03-20
- made static and compress work when not an express server

## [0.6.1] - 2017-03-22
### Enhanced
- colorized docs

## [0.5.2] - 2017-02-08
### update ack-node which update ack-path which had breaking changes

## [0.5.1] - 2017-02-03
### Added
- Built-in Testing

## [0.5.0] - 2017-02-01
### Added
- Documentation
- webapp.noRobots


## [0.4.0] - 2017-01-23
### Updated Dependencies

## [0.3.3] - 2017-01-10
### Adjusted
- tab stops

## [0.3.2] - 2017-01-04
### Enhanced
- site apps now have .beforeStart() and .afterStart() methods

## [0.3.0] - 2016-11-07
### Added
- host.web().startOnePort()
### Removed
- reqrouting (on request js files that match folder paths)

## [0.2.4] - 2016-11-07
### Fixed
- api no longer auto parses client input

js/web.js
"use strict";
exports.__esModule = true;
var ack_node_1 = require("ack-node");
var reqresRes = require("ack-node/js/modules/reqres/res");
var vhost = require("vhost");
var https = require("https");
var webapp_1 = require("./webapp");
var tls = require("tls");
var express = require("express");
var routers = ack_node_1.ackX.router();
var startTestingRoutes = require('./startTestingRoutes.function');
var testSite = require('./testRoutes.function');
/**
  request object CLASS
  @scope: isClearRequires, consoleAll
*/
var web = function web() {
    this.router = routers;
    this.data = {};
    this.data.portStruct = this.data.portStruct || {};
    this.data.routers = {};
    this.data.isClearRequires = this.data.isClearRequires || this.isProductionMode;
    this.data.consoleAll = this.data.consoleAll;
    this.data.consoleNonProductionErrors = true;
};
web.app = webapp_1.webapp.webapp;
web.prototype.consoleAll = function (logger) {
    this.data.routers.consoleAll = webapp_1.webapp.consoleAll(logger);
    return this;
};
web.prototype.getRunningMode = function () {
    if (this.data.isProductionMode)
        return 'PRODUCTION';
    return process.env.NODE_ENV;
};
web.prototype.isProductionMode = function (yesNo) {
    if (yesNo != null) {
        this.data.isProductionMode = yesNo;
        return yesNo;
    }
    if (this.data.isProductionMode != null)
        return this.data.isProductionMode;
    return (this.getRunningMode() || 'development').toLowerCase() == 'production';
};
//web.prototype.routers = webapp
web.prototype.getExpress = function () {
    return express;
};
web.prototype.registerApp = function (app, portOrArray, host, sslOps) {
    if (portOrArray && portOrArray.constructor == Array) {
        for (var x = 0; x < portOrArray.length; ++x) {
            this.registerOneApp(app, portOrArray[x], host, sslOps);
        }
    }
    else {
        this.registerOneApp(app, portOrArray, host, sslOps);
    }
    return this;
};
/**
  @app - any connect/express like app with use/get/put/post/all methods
  @port - single port or multiple port array
  @host - domain host name or array of host names
  @sslOps - {SNICallback, cert, key}
*/
web.prototype.registerOneApp = function (app, port, host, sslOps) {
    var portStruct = this.paramPortStruct(port);
    portStruct.appArray.push(app); //catalog app
    /* add events */
    app.beforeStart = function (method) {
        if (method == 'start') {
            return app.beforeStart.memory.forEach(function (callback) { return callback(); });
        }
        app.beforeStart.memory.push(method);
    };
    app.beforeStart.memory = [];
    app.afterStart = function (method) {
        if (method == 'start') {
            return app.afterStart.memory.forEach(function (callback) { return callback(); });
        }
        app.afterStart.memory.push(method);
    };
    app.afterStart.memory = [];
    /* end: add events */
    if (app.isProductionMode) {
        app.isProductionMode(this.isProductionMode()); //pass production mode status along
    }
    app.use(reqresnext(this)); //every request is touched by web processor
    //when engaged by request, ensure default app does not run by setting host-name
    app.use(function (req, res, next) {
        res.ackHostHosted = host || port; //let the main app know a vhost||port picked up the request and not to run default app
        next();
    });
    if (host) {
        if (host.constructor == String) { //cast host string to host array. single-host versus !multi-host-array
            host = [host];
        }
        host.forEach(function (hostName, i) {
            portStruct.hostArray.push(hostName);
            var vRoute = vhost(hostName, app);
            portStruct.rootApp.use(vRoute); //request processor(analyze host-name for processing
        });
    }
    if (sslOps) {
        if (host) {
            portStruct.multiSslArray.push({ hostNameArray: host, sslOps: sslOps });
        }
        if (!portStruct.sslOps) {
            portStruct.sslOps = sslOps;
        }
    }
    return this;
};
/** returns connect-package app. args: portNumber, virtualHostName (does not pre-load clientInput)
  @port - single port or multiple port array
  @hostOrOptions - domain host name or array of host names || @options
  @options{
    cert:String - for ssl
    key:String - for ssl
    SNICallback:Function - for ssl
  }
*/
web.prototype.host = function (port, hostName, options) {
    var app = webapp_1.host(); //connect app WITH addons
    this.registerApp(app, port, hostName, options);
    return app;
};
/** returns express app. Defaults include request timeout, force strict-paths, and 404 is closing midware
  @port - single port or multiple port array
  @hostOrOptions - domain host name or array of host names || @options
  @options{
    timeout:Number = 10000
    cert:String - for ssl
    key:String - for ssl
    SNICallback:Function - for ssl
  }
*/
web.prototype.api = function (port, hostOrOptions, options) {
    if (hostOrOptions && hostOrOptions.constructor == Object) {
        options = hostOrOptions;
        hostOrOptions = null;
    }
    else {
        options = options || {};
    }
    var app = new webapp_1.webapp.reExpress().strictPaths();
    if (options.timeout || options.timeout == null) {
        app.use(routers.timeout(options.timeout || 10000));
    }
    this.registerApp(app, port, hostOrOptions, options);
    app.beforeStart(function () {
        app.use(routers.notFound());
    });
    return app;
};
/** same as api but always pre-loads clientInput. Possibly deprecated */
web.prototype.website = function (port, host, sslOps) {
    var app = new webapp_1.webapp.webapp()
        .strictPaths()
        .preloadClientInput()
        .timeout(30000);
    this.registerApp(app, port, host, sslOps);
    app.beforeStart(function () {
        app.use(routers.notFound());
    });
    return app;
};
/** Returns object of test results {passing:[],failing:[]}
  @options{
    port   : Number = all-ports
    host   : String = localhost
    method : String (GET|POST|PUT|DELETE)
    logTo  : Object = console.log - not yet implemented
  }
*/
web.prototype.startAndTest = function (options) {
    return startTestingRoutes(this, this.data.portStruct, options);
};
/** see startAndTest */
web.prototype.test = function (options) {
    if (options === void 0) { options = {}; }
    this.eachApp(function (app, port) {
        var ops = Object.assign({}, options);
        ops.port = port;
        testSite(app, ops);
    }, options);
};
/** see startAndTest */
web.prototype.testApps = function (apps, options) {
    for (var x = 0; x < apps.length; ++x) {
        testSite(apps[x], options);
    }
};
web.prototype.getAppsByPort = function (port) {
    return this.data.portStruct[port].appArray;
};
web.prototype.eachApp = function (callback, options) {
    if (options === void 0) { options = {}; }
    var ports = options.port ? [options.port] : Object.keys(this.data.portStruct);
    for (var portIndex = 0; portIndex < ports.length; ++portIndex) {
        var port = ports[portIndex];
        for (var appIndex = 0; appIndex < this.data.portStruct[port].appArray.length; ++appIndex) {
            callback(this.data.portStruct[port].appArray[appIndex], port);
        }
    }
};
/** as soon as one port is available to start, it is started */
web.prototype.startOnePort = function (onPortStart, options) {
    if (options === void 0) { options = {}; }
    options.portStartCount = 1;
    return this.start(onPortStart, options);
};
//each(portNum, server, rootApp, portStruct) = as each port is started
web.prototype.start = function (each, options) {
    var _this = this;
    if (options === void 0) { options = {}; }
    each = each || function () { }; //function to call with each starting port
    var portNumArr = Object.keys(this.data.portStruct);
    var promise = ack_node_1.ackX.promise().bind(this);
    if (options.portStartCount) {
        var pos_1 = 0;
        var rotator_1 = function (callback) {
            var port = Number(portNumArr[pos_1]);
            if (pos_1 >= portNumArr.length) {
                //throw new Error('ran out of ports')
                callback(new Error('Ran out of ports to try to start server with. Ports Tried:' + portNumArr.join(',')));
            }
            return _this.startPort(port)
                .then(function (server) {
                callback(null, [server]);
            })["catch"]('EADDRINUSE', function (e) {
                console.log('\x1b[34mack-host:\x1b[0m \x1b[33mport ' + port + ' is in use\x1b[0m');
                ++pos_1;
                return rotator_1(callback);
            });
        };
        promise = promise.callback(rotator_1);
    }
    else {
        promise = promise.map(portNumArr, this.startPort);
    }
    return promise
        .map(function (array) {
        //console.log('array', array[0])
        var portNum = array[0], portStruct = array[1];
        each(portNum, portStruct.server, portStruct.rootApp, portStruct);
    })
        .set(this.data.portStruct);
};
web.prototype.startPort = function (portNum) {
    portNum = Number(portNum);
    var portStruct = this.data.portStruct[portNum];
    //ports default app
    if (portStruct.appArray.length) {
        var defApp = portStruct.appArray[0];
        portStruct.rootApp.use(function (req, res, next) {
            if (!res.ackHostHosted) { //indicator of vhost processing, not present, run Default-App
                defApp(req, res, next);
            }
            else {
                next();
            }
        });
    }
    portStruct.appArray.forEach(function (app, i) {
        app.beforeStart('start');
    });
    if (!this.isProductionMode() && this.data.consoleNonProductionErrors) {
        //console.log('--Non-Production-Mode-Detected (mode='+this.getRunningMode()+'): All request errors will be console logged.')
        portStruct.rootApp.use(routers.consoleNonProductionErrors());
    }
    //request closing
    portStruct.rootApp.use(reqresRes.geterrhan());
    portStruct.rootApp.use(function (req, res, next) {
        if (!res.closed && !res._headerSent) {
            ack_node_1.ackX.reqres(req, res).abort();
        }
    });
    var ssl = portStruct.sslOps && portStruct.sslOps.cert && portStruct.sslOps.key;
    var tApp;
    if (ssl) {
        var sslOps = portStruct.sslOps;
        /* DYNAMIC SSL */
        var isSniMode = portStruct.multiSslArray.length > 1;
        if (isSniMode) {
            if (portStruct.sslOps.SNICallback) { //already exist?
                sslOps = {}; //ensure sslOps.SNICallback won't be overwritten on original scope. If two ports start a server, the last one would always be the SNI responder for all request (bad)
                for (var key in portStruct.sslOps) {
                    sslOps[key] = portStruct.sslOps[key];
                }
            }
            sslOps.SNICallback = getSniCallbackForPort(portNum, this);
        }
        /* END: DYNAMIC SSL */
        tApp = https.createServer(sslOps, portStruct.rootApp);
    }
    else {
        tApp = portStruct.rootApp;
    }
    return ack_node_1.ackX.promise().bind(tApp)
        .callback(function (callback) {
        //portStruct.server = tApp.listen(portNum,'0.0.0.0',function(err,data){
        portStruct.server = tApp.listen(portNum, '0.0.0.0', undefined, callback);
        portStruct.server.on('error', function (e) {
            callback(e);
        });
    }) //start the http.listen protcol on a certain port, with IPV4, and with a start-up-complete callback
        .then(function () {
        //tell all apps we have started
        portStruct.appArray.forEach(function (app, i) {
            app.events.emit('start', portStruct.server);
            app.afterStart('start');
        });
    })
        .set([portNum, portStruct]);
};
function getSniCallbackForPort(portNum, hostApp) {
    return function (domain, callback) {
        //this.server._connectionKey//server identifiable string (always handy)
        /* !!! keep!  This is a test to ensure the same portNum SNICallback is for the right server. If SSL issues occur,
        if(this.server && this.server._connectionKey){
          var reqPort = this.server._connectionKey.split(':').pop()
          reqPort = Number( reqPort )
          if(reqPort!=portNum){
            console.log('SNICallback crash detected. '+reqPort+' & '+portNum)
            throw new Error('SNICallback crash detected. '+reqPort+' & '+portNum)
          }
        }
        */
        var portStruct = hostApp.data.portStruct[portNum];
        for (var msa = 0; msa < portStruct.multiSslArray.length; ++msa) {
            var hostConfig = portStruct.multiSslArray[msa];
            for (var hna = 0; hna < hostConfig.hostNameArray.length; ++hna) {
                if (hostConfig.hostNameArray[hna].toLowerCase() == domain.toLowerCase()) {
                    var ssl = tls.createSecureContext(hostConfig.sslOps).context;
                    return callback(null, ssl);
                }
            }
        }
        //var ssl = crypto.createCredentials(hostConfig.sslOps).context
        var ssl = tls.createSecureContext(portStruct.sslOps).context;
        callback(null, ssl);
    };
}
web.prototype.stop = function (each) {
    var ps = this.data.portStruct;
    var keys = Object.keys(ps);
    var stopped = keys.length;
    return ack_node_1.ackX.promise()
        .next(function (next) {
        function closer() {
            if (stopped == 0) {
                next();
            }
        }
        keys.forEach(function (v, i) {
            if (!ps[v].server) {
                --stopped;
                if (!stopped) {
                    next();
                }
                return; //nothing to stop here
            }
            ps[v].appArray.forEach(function (v, i) {
                v.events.emit('stop', v.server);
            });
            ps[v].server.close(function () {
                --stopped;
                var server = ps[v].server;
                delete ps[v].server;
                if (each)
                    each(v);
                closer();
            });
        });
    });
};
web.prototype.getPortCount = function () {
    return Object.keys(this.data.portStruct).length;
};
/** returns portnumber of port that is still on */
web.prototype.isOn = function () {
    var portNum, i, portKeys = Object.keys(this.data.portStruct);
    if (!this.data.portStruct || !portKeys.length)
        return 0;
    for (i = 0; i < portKeys.length; ++i) {
        portNum = portKeys[i];
        if (this.data.portStruct[portNum].server) {
            return portNum;
        }
    }
    return false;
};
web.prototype.paramPortStruct = function (port) {
    if (this.data.portStruct[port])
        return this.data.portStruct[port];
    this.data.portStruct[port] = {
        appArray: [], hostArray: [], sslOps: null, multiSslArray: [],
        rootApp: webapp_1.host() //.use(this.getreqhan())
    };
    return this.data.portStruct[port];
};
web.prototype.dropPorts = function () {
    this.data.portStruct = {};
};
function reqresnext(iWeb) {
    return function (req, res, next) {
        return ack_node_1.ackX.promise().map(iWeb.data.routers, function (value, key, len) {
            return ack_node_1.ackX.promise().callback(function (callback) {
                return value(req, res, callback);
            });
        })
            .set().then(next)["catch"](next);
    };
}
exports.module = new web();
//module.exports.Class = web
